### PR TITLE
feat(executor): implement stack overflow protection

### LINKED ISSUE #2079: Unexpected behavior
### Current State
It seems the program is in an infinite loop and it doesn't output results

### Expected

Raise an exception indicating stack overflow.

## Environment

-  x86\_64 
- Operating system: Ubuntu 20.04

The following information is optional. Please provide them only if you have built from source.
- C++ Compiler version: clang-14
- CMake version: 3.21.6
- CMake flags:  `-DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_AOT_RUNTIME=ON`
- Boost version: libboost1.71-dev:amd64

- build commands
  ```
  export CC=/usr/bin/clang-14
  export CXX=/usr/bin/clang++-14
  rm -rf build ; mkdir build ; 
  cd build ; cmake -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_AOT_RUNTIME=ON .. ; make -j ; cd ..
  ```

## Steps to Reproduce
[wasmedge_timeout.zip](https://github.com/WasmEdge/WasmEdge/files/9997777/wasmedge_timeout.zip)

1. build the wasmedge
2. run the test case by `build/tools/wasmedge/wasmedge --reactor <test_case_name> to_test `

### PR DESCRIPTION
**Pull request by criticic**
Add configurable call stack depth limit to prevent infinite recursion and stack overflow issues. Fixes #2079 and #4410.

Changes:
- Add CallStackExhausted error code (0x041A) in execution phase
- Add MaxCallDepth configuration option (default: 10000) to RuntimeConfigure
- Add stack depth checks before pushing frames in enterFunction()
  - Checks added for host functions, compiled functions, and native functions
- Implement assert_exhaustion handler in spec tests
- Add C API functions: WasmEdge_ConfigureSetMaxCallDepth() and WasmEdge_ConfigureGetMaxCallDepth()
- Add comprehensive unit tests for stack overflow scenarios

### CODE REVIEW

File: test/spec/spectest.cpp [hunk 1]
```diff
@@ -921,7 +921,21 @@ void SpecTest::run(std::string_view Proposal, std::string_view UnitName) {
...
+        const simdjson::dom::object &Action = Cmd["action"];
+        const uint64_t LineNumber = Cmd["line"];
+        const auto ModName = GetModuleName(Action);
+        const std::string_view Field = Action["field"];
+        simdjson::dom::array Args = Action["args"];
+        const auto Params = parseValueList(Args);
+
+        if (auto Res = onInvoke(ModName, std::string(Field), Params.first,
+                                Params.second)) {
+          EXPECT_NE(LineNumber, LineNumber);
```
[the remark this answers is not in the document]
Comment (criticic) [reply test/spec/spectest.cpp:933]: Not relevant! This is just used to report line numbers, and is the same way throughout the file.

**Comment by hydai** [comment]
Hi @criticic,  
Can you confirm if these tests are runnable? I think they may encounter timeout errors in several CI jobs.

**Comment by q82419** [comment]
In my opinion, I believe this implementation surely can prevent from the runtime overflow becuase it strickly limited the calling frame size.
But I'm still considering about the appropriate value of limited calling frame size. Maybe we should study the details in [WASM tool conventions](https://github.com/WebAssembly/tool-conventions).
@criticic Your PR is meaningful but we should keep discussing with this. Thanks!

**Comment by criticic** [comment]
> appropriate value of limited calling frame size

I could not find anything in the spec about what the default call stack size should be. I'll also look at other implementations and see what they have done regarding this.
